pkg/nuclide-swift/lib/taskrunner/SwiftPMTaskRunner.js

```diff
@@ -0,0 +1,204 @@
+import type {Directory, Task, TaskMetadata} from '../../../nuclide-task-runner/lib/types';
+import type {Level, Message} from '../../../nuclide-console/lib/types';
+import type {SwiftPMTaskRunnerStoreState} from './SwiftPMTaskRunnerStoreState';
+
+import invariant from 'assert';
+import {Observable, Subject} from 'rxjs';
+import {Dispatcher} from 'flux';
+import {CompositeDisposable} from 'atom';
+import {React} from 'react-for-atom';
+import fsPromise from '../../../commons-node/fsPromise';
+import {DisposableSubscription} from '../../../commons-node/stream';
+import {observeProcess, safeSpawn} from '../../../commons-node/process';
+import {observableToTaskInfo} from '../../../commons-node/observableToTaskInfo';
+import SwiftPMTaskRunnerStore from './SwiftPMTaskRunnerStore';
+import SwiftPMTaskRunnerActions from './SwiftPMTaskRunnerActions';
+import {buildCommand, testCommand} from './SwiftPMTaskRunnerCommands';
+import {
+  SwiftPMTaskRunnerBuildTaskMetadata,
+  SwiftPMTaskRunnerTestTaskMetadata,
+  SwiftPMTaskRunnerTaskMetadata,
+} from './SwiftPMTaskRunnerTaskMetadata';
+import SwiftPMTaskRunnerToolbar from './toolbar/SwiftPMTaskRunnerToolbar';
+import {SwiftIcon} from '../ui/SwiftIcon';
+
+/**
+ * nuclide-swift makes use of the Flux design pattern. The SwiftPMTaskRunner is
+ * responsible for kicking off SwiftPM tasks such as building a package. How it
+ * builds the package is determined by the state of the
+ * SwiftPMTaskRunnerToolbar -- the path to the package, whether a build path is
+ * specified, etc. -- and that state is maintained by the
+ * SwiftPMTaskRunnerStore. Updates to the toolbar UI options trigger actions,
+ * defined in SwiftPMTaskRunnerActions, which update the state of the store.
+ * Actions are routed to the store via a Flux.Dispatcher (instantiated by
+ * SwiftPMTaskRunner).
+ */
+type SwiftPMTaskRunnerFlux = {
+  store: SwiftPMTaskRunnerStore,
+  actions: SwiftPMTaskRunnerActions,
+};
+
+/**
+ * The primary controller for spawning SwiftPM tasks, such as building a
+ * package, or running its tests. This class conforms to Nuclide's TaskRunner
+ * interface.
+ */
+export class SwiftPMTaskRunner {
+  id: string;
+  name: string;
+  _disposables: CompositeDisposable;
+  _initialState: ?SwiftPMTaskRunnerStoreState;
+  _flux: ?SwiftPMTaskRunnerFlux;
+  _taskList: Observable<Array<TaskMetadata>>;
+  _outputMessages: Subject<Message>;
+
+  constructor(initialState: ?SwiftPMTaskRunnerStoreState) {
+    this.id = 'swiftpm';
+    this.name = 'Swift';
+    this._initialState = initialState;
+    this._disposables = new CompositeDisposable();
+    this._outputMessages = new Subject();
+    this._disposables.add(this._outputMessages);
+  }
+
+  dispose(): void {
+    this._disposables.dispose();
+  }
+
+  serialize(): SwiftPMTaskRunnerStoreState {
+    return this._getFlux().store.serialize();
+  }
+
+  getExtraUi(): ReactClass<any> {
+    const {store, actions} = this._getFlux();
+    return class ExtraUi extends React.Component {
```

Comment:
Hm, I think I'm actually calling this several times. I guess I should change that!